connectors/cli.py

```diff
@@ -95,34 +96,17 @@ async def _start_service(config, loop):
         for sig in (signal.SIGINT, signal.SIGTERM):
             loop.remove_signal_handler(sig)
 
-    services = [SyncService(config), JobCleanUpService(config)]
+    multiservice = MultiService([SyncService(config), JobCleanUpService(config)])
     for sig in (signal.SIGINT, signal.SIGTERM):
+        loop.add_signal_handler(sig, functools.partial(multiservice.shutdown, sig.name))
```

Comment:
Should it be though? It requires access to `loop` that is running and `cli.py` already has it, so I'd just leave the code as is.

Reply:
The loop is available everywhere once your app starts, with `asyncio.get_running_loop` https://docs.python.org/3/library/asyncio-eventloop.html#asyncio.get_running_loop

Reply:
I know, but it does not feel like adding signal handlers should leave in multiple places of the project.

Reply:
Indeed, this is why I was thinking about this.

if we look at this code here: https://github.com/elastic/connectors-python/blob/41a54228b0181f4a538707a58572e4f16beec5cd/connectors/cli.py#L89-L101

There's a pattern where we give a service the handling of signals, and that code is duplicated. If the Service class internalizes it, the block could become:

```python
    preflight = PreflightCheck(config)
    with preflight.handle_shutdown():
        if not await preflight.run():
            return -1

    multiservice = MultiService(SyncService(config), JobCleanUpService(config))
    with multiservice.handle_shutdown():
        return await multiservice.run()
```

and simplified with:

```python

    with PreflightCheck(config) as service:
        if not await service.run():
            return -1

    with MultiService(SyncService(config), JobCleanUpService(config)) as service:
        return await service.run()
```

but that's just food for thoughts for another PR, not required!

Reply:
My main concern is that I'm just not 100% sure what happens, if you add multiple signal handlers at the same time to different services.

Imagine, for example, by mistake the following logic is introduced:

```
multiservice = MultiService(SyncService(config), JobCleanUpService(config))
with multiservice.handle_shutdown():
     return await multiservice.run()

# Somewhere deep inside the framework
for sig in (signal.SIGINT, signal.SIGTERM): 
     loop.add_signal_handler(sig, functools.partial(logger.debug, "something")) 

```

Will in this case both be handled or only the last handler called?

Reply:
no you should not have mutiple services handling the same signals. This is why implementing it at the base class level solves it -- the base class can deal with this unicity and make sure the handler cannot be registered twice. The current code is doing this manually.

Reply:
> PreflightCheck is not a service though

it's implemented as a service right now (same APIs)

> To me it becomes more confusing - now it's all in one file, with the change it'll be in different namespaces.

I think it boils down to:

-  the fact that the code that registers and unregisters signals is currently duplicated because we have a two-phase startup. 
- defining the responsibility of each class. If the `Service` class is the one responsible for a shutdown, it in its scope 

Maybe the confusion is around what `MultiService` is. Maybe it's not a service and should have its own implementation where it takes care of the preflight check and starting services

Reply:
> Maybe the confusion is around what MultiService is. Maybe it's not a service and should have its own implementation where it takes care of the preflight check and starting services

Hmmm that makes sense, I'll see how things can move around so that there's only one registration in the whole code!